llvm/lib/CodeGen/Yk/FixStackmapsSpillReloads.cpp

```diff
@@ -0,0 +1,211 @@
+//   ...
+//   call foo($rsi, $rdi)
+//   movrr $rsi, $rbx
+//   movrm $rdi, $rbp, -8
+//   STACKMAP $rsi, $rdi
+//
+// In order to pass arguments to foo, the register allocator had to spill the
+// values in $rdi and $rsi into another register or onto the stack before the
+// call. Then immediately after the call it inserted instructions to reload
+// the spilled values back into the original registers. Since during
+// deoptimisation we return to immediately after the call, the stackmap is now
+// tracking the wrong values, e.g. in this case $rdi and $rsi instead of the
+// spill locations.
+//
+// This pass interates over all basic blocks, finds spill reloads inserted
+// inbetween a call and stackmap, replaces the stackmap operands with the
+// spill reloads, and then moves the stackmap instruction up just below the
+// call.
+//===----------------------------------------------------------------------===//
+
+#include "llvm/CodeGen/MachineBasicBlock.h"
+#include "llvm/CodeGen/MachineFrameInfo.h"
+#include "llvm/CodeGen/MachineFunction.h"
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineInstr.h"
+#include "llvm/CodeGen/MachineInstrBuilder.h"
+#include "llvm/CodeGen/MachineOperand.h"
+#include "llvm/CodeGen/Passes.h"
+#include "llvm/CodeGen/StackMaps.h"
+#include "llvm/CodeGen/TargetInstrInfo.h"
+#include "llvm/IR/DebugLoc.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Support/Debug.h"
+
+using namespace llvm;
+
+#define DEBUG_TYPE "fix-stackmaps-spill-reloads"
+
+namespace {
+
+class FixStackmapsSpillReloads : public MachineFunctionPass {
+public:
+  static char ID;
+
+  FixStackmapsSpillReloads() : MachineFunctionPass(ID) {
+    initializeFixStackmapsSpillReloadsPass(*PassRegistry::getPassRegistry());
+  }
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.setPreservesCFG();
+    MachineFunctionPass::getAnalysisUsage(AU);
+  }
+
+  StringRef getPassName() const override {
+    return "Stackmaps Fix Post RegAlloc Pass";
+  }
+
+  bool runOnMachineFunction(MachineFunction &MF) override;
+};
+
+} // namespace
+
+char FixStackmapsSpillReloads::ID = 0;
+char &llvm::FixStackmapsSpillReloadsID = FixStackmapsSpillReloads::ID;
+
+INITIALIZE_PASS_BEGIN(FixStackmapsSpillReloads, DEBUG_TYPE, "Fixup Stackmap Spills",
+                      false, false)
+INITIALIZE_PASS_END(FixStackmapsSpillReloads, DEBUG_TYPE, "Fixup Stackmap Spills",
+                    false, false)
+
+
+bool FixStackmapsSpillReloads::runOnMachineFunction(MachineFunction &MF) {
+  bool Changed = false;
+  const TargetInstrInfo *TII = MF.getSubtarget().getInstrInfo();
+  for (MachineBasicBlock &MBB : MF) {
+    bool Collect = false;
+    std::set<MachineInstr *> Erased;
+    MachineInstr *LastCall = nullptr;
+    std::map<Register, MachineInstr *> Spills;
+    for (MachineInstr &MI : MBB) {
+      if (MI.isCall() && !MI.isInlineAsm()) {
+        // YKFIXME: Do we need to check for intrinsics here or have they been
```

Comment:
Can we add an `assert` to check that we don't see them here?

Reply:
If I knew how to check for intrinsics I would have added that check to the if-statement. But there's no (reliable) way to check if a `MachineInstr` is an intrinsic call. At least none that I know off. I experimented with various things, e.g. `isMCSymbol` but that one wasn't always true for all non-intrinsic function calls.